### DISCUSSION TITLE
Ssh clone works in Git bash but not in WSL2 git

### DISCUSSION THREAD

**Discussion by bluesherpa** [community/community#22218] [New to GitHub]
<p>Could someone help me figure out how to get git clone over ssh working in Windows Subsystem for Linux (WSL2) with Ubuntu version 20.04.1 LTS? Testing the ssh connection or attempting to clone a repo over ssh both fail with the error message “Received disconnect from 192.30.255.112 port 22:11: Bye Bye” and “fatal: Could not read from remote repository.”</p>
<p>Here are the commands I’m using:</p>
<ul>
<li>ssh -T git <span class="mention">@github.com</span>
</li>
<li>git clone git <span class="mention">@github.com</span>:bluesherpa/array_sort.git ./</li>
</ul>
<p>Note: space between git and @ added so I can submit this post.</p>
<p>Performing the clone over HTTPS works:</p>
<ul>
<li>git clone <a href="https://github.com/bluesherpa/array_sort.git" rel="noopener nofollow ugc">https://github.com/bluesherpa/array_sort.git</a> ./</li>
</ul>
<p>I run these commands to setup SSH from within an Ubuntu bash terminal:</p>
<ul>
<li>rm -rf ~/.ssh</li>
<li>pkill ssh-agent</li>
<li>ssh-keygen -t rsa -C “”</li>
<li>eval <code>ssh-agent -s</code>
</li>
<li>ssh-add -D</li>
<li>ssh-add ~/.ssh/id_rsa</li>
</ul>
<p>I realize there are some extra steps there. I use pkill to stop the ssh-agent process and make sure that’s not causing problems, same with using “ssh-add -D” to make sure there are no certificates.</p>
<p>I copy the content of ~/.ssh/id_rsa.pub and add the corresponding SSH key in my GitHub account.</p>
<p>I try the testing the ssh connection and cloning a repo after everything is setup with no luck.</p>
<p>None of this works from Ubuntu under WSL2, but to cross-reference this, I removed ~/.ssh and killed the ssh-agent and tried everything from scratch using Git Bash - it works from there.</p>
<p>I tried checking which version of git is installed in Ubuntu. It was an old version, so I upgraded it so now git.exe from the Git Bash Windows installation and the Ubuntu git are the same version.</p>
<p>I tried using the version of git that comes with Git Bash under Ubuntu with the command:</p>
<ul>
<li>git.exe clone <a href="https://github.com/bluesherpa/array_sort.git" rel="noopener nofollow ugc">https://github.com/bluesherpa/array_sort.git</a> ./</li>
</ul>
<p>That works.</p>
<p>Basically, I can get everything working under Git Bash, but not Ubuntu and I’m not sure why. I even tried copying the .ssh directory from my home directory in Windows to my home directory in Ubuntu.</p>
<p>I tried uninstalling/reinstalling Git Bash and trying different options for the credential helper, including using the Git Credential Manager Core, the Git Credential Manager, and none. None of that made a difference, tho I found it interesting that running ‘git clone’ under Ubuntu would trigger a popup menu from Git Credential Manager Core, which doesn’t work and does a fail over to basic authentication at the command prompt.</p>
<p>I have tried all of this using an Ubuntu Bash Shell from the Windows Start menu, an Ubuntu Bash Shell from Windows Terminal, and an Ubuntu Bash Shell “Terminal” from Visual Code.</p>
<p>Another thing I tried was cross-referencing this with using a phrase and not using a phrase when generating the keys.</p>
<p>Finally… I notice that when I try this from Git Bash, I get prompted to create a new entry in known_hosts. This does not happen when I try with Git in Ubuntu. I tried just copying the known_hosts file to ~/.ssh but that didn’t seem to help.</p>
<p>I’m not sure if this information is being cached anywhere, but GitHub seems respond pretty quickly every time I delete and recreate the keys from within Git Bash.</p>
<p>Here are the documents I’ve been trying to follow:</p>
<ul>
<li>https:  //  docs.github.  com /en/free-pro-team@latest/github/authenticating-to-github/connecting-to-github-with-ssh</li>
<li>https:  //  docs.microsoft.  com /en-us/windows/wsl/tutorials/wsl-git</li>
</ul>
<p>Sorry if this is too much detail, but I really do appreciate your help.</p>
<p>-Ed</p>

**Comment by bluesherpa** [comment]
<p>I just thought of something else to add that I’ve done while troubleshooting this…  I went into the Windows Credential Manager &gt; Windows Credentials &gt; Generic Credentials… and made sure to remove anything associated with GitHub.</p>

**Comment by bluesherpa** [comment]
<p>I had another idea to cross-reference this using Debian instead of Ubuntu under Windows 10 (WSL2)… Everything is working within Debian. Any ideas on why Ubuntu isn’t working?</p>
<p>-Ed</p>

**Comment by bluesherpa** [comment]
<p>I decided this must be something specific to the copy of Ubuntu 20.04 LTS available in the Microsoft App store for installation on Windows 10 because I don’t run into any of these problems using the 18.04 LTS version of Ubuntu available from the Microsoft App store or the Debian installation.</p>
<p>Steps to reproduce are pretty straight forward…</p>
<ol>
<li>Uninstall Ubuntu</li>
<li>Remove SSH key defined in GitLab account</li>
<li>Install Ubuntu</li>
<li>Launch Ubuntu and specify the name and password for the default account.</li>
<li>Run "ssh-keygen -t rsa -C “” and accept defaults</li>
<li>Copy ~/.ssh/id_rsa.pub and define a new SSH key in the GitHub account</li>
<li>Try: ssh -T <a href="mailto:git@github.com">git@github.com</a>
</li>
<li>Try: cloning an SSH repo</li>
</ol>
<p>This works with Ubuntu 18.04 LTS but not with 20.04 LTS.</p>
<p>This all works for Ubuntu 18.04 LTS but not for 20.04 LTS.</p>

**Comment by darkRaspberry** [comment, marked as answer]
<p>I used SSH in 20.04 LTS 18.04 LTS and 20.04.1 LTS. It all works in good way. I can’t understand.</p>
<p>One more thing whats this <code>./</code> thing.<br>
I guess this command is enough for a clone using ssh.<br>
<code>git clone git@github.com:username/repo.git</code></p>
<aside class="quote no-group" data-username="bluesherpa" data-post="1" data-topic="147156">
<div class="title">
<div class="quote-controls"></div>
 bluesherpa:</div>
<blockquote>
<p>git clone git <span class="mention">@github.com</span>:bluesherpa/array_sort.git ./</p>
</blockquote>
</aside>
<p>Now look here. You added a space between <code>git</code> and <code>@github.com</code> which is not required. and last <code>./</code> this is also not required.<br>
Can you use this and confirm whats showing in the WSL with a screenshot</p>
<p><code>git clone git@github.com:bluesherpa/array_sort.git</code></p>

**Comment by darkRaspberry** [comment]
<aside class="quote no-group" data-username="bluesherpa" data-post="2" data-topic="147156" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 bluesherpa:</div>
<blockquote>
<p>I just thought of something else to add that I’ve done while troubleshooting this… I went into the Windows Credential Manager &gt; Windows Credentials &gt; Generic Credentials… and made sure to remove anything associated with GitHub.</p>
</blockquote>
</aside>
<p>WSL is nested virtualisation There is no need for these. both don’t share same place for credential.  One is Linux and other is windows. They are far beyond.</p>

**Comment by bluesherpa** [comment]
<p>Hi DarkRaspberry,</p>
<p>Regarding the spaces, I mentioned in the first post that I had to add those in order to submit the post. I didn’t have the spaces while testing and trying to get this to work. My account is still considered New since I haven’t posted much here, so there was I was limited in the number of links I could include and had to use formatting with spaces as a work around solution.</p>
<p>What you’re saying about the Windows Credentials makes sense. One of the help articles I referenced includes those steps. Apparently the Git Credential Manager is being used in some configurations. I agree it shouldn’t be necessary, but mentioned it in case it’s part of the problem.</p>
<p>It’s great to hear it worked for you on on 20.04 LTS. Did you use the steps I outlined? I am running into the problem with a new installation of 20.04 LTS and the steps I provided.</p>

**Comment by bluesherpa** [comment]
<p>Hi darkRaspberry!</p>
<p>I did more testing today. I have no idea what’s causing the issue, but I created a bash script and a set of key files to make sure I am testing everything consistently in each environment.</p>
<p>Everything works in fresh installations of Ubuntu 18.04 LTS and Debian.</p>
<p>The problem shows up in a fresh installation of Ubuntu 20.04 LTS.</p>
<p>The problem also shows up if I take an instance of Ubuntu 18.04 LTS and upgrade it to Ubuntu 20.04 LTS.</p>
<p>I noticed the upgrade results in going from OpenSSH 7.6p1 to OpenSSH 8.2p1 and from OpenSSL 1.0.2n to OpenSSL 1.1.1f, so I did an experiment of doing a fresh installation of Ubuntu 18.04 LTS and only upgrading OpenSSH and OpenSSL. The problem did not show up. That convinces me the problem has something to do with Ubuntu 20.04 LTS, but I’m not sure what it is, especially since everything works on your copy of it.</p>
<p>-Ed</p>

**Comment by bluesherpa** [comment]
<p>Marking this closed. Another friend confirmed that they run into the same problem from their computer when doing a fresh installation of Ubuntu 20.04 LTS.</p>

**Reply by FrustrumCode** [reply]
This can't still be an issue... but I'm getting asked for passphrase on every push (latest ubuntu on wsl2, really not feeling like rolling back and would like to fix).

**Reply by FrustrumCode** [reply]
Ancient thread, but here is the fix I found that worked (don't fully understand it since I live in emacs more than linux)...

`sudo apt install keychain`

Then add the following line to your ~/.bashrc file…

`eval ``keychain --eval --agents ssh id_rsa`

Then I type bash in the ubuntu terminal.  It asks for the pw.  Then run emacs from there and am no longer asked for the passphrase everytime I push.

**Comment by GebTorte** [comment]
Hi, for me simply exchanging 
`git clone ...` 
with 
`git.exe clone ... `
worked magically (Ubuntu 24.04 WSL2 in 2025).
